### PR TITLE
Less cloning in `BlobDataProvider`

### PR DESCRIPTION
**Pull request by robertbastian**
This changes `BlobDataProvider`'s cart to `Option<(A)rc<Box<[u8]>>`. Previously we used `Option<(A)rc<[u8]>>`, which was expensive to create from `Box[u8]` and `Vec<u8>`, as it required an allocation the size of the buffer. The constructor now accepts `Box<[u8]>`.

The cart type is now defined in the `icu_provider` crate. This is so that the same cart is used by `BlobDataProvider` and `DataResponse`, and that future changes (like erasing the type) are easier. This also allows removing the `RcWrap` type, as the cart type now changes on the `sync` feature.

~Lastly, the FFI `BlobDataProvider` constructor was changed to use `BlobDataProvider::from_static_bytes`, which is really just `from_borrowed_bytes` over FFI. This decreases the code size if the tiny example from 27000 to 26880.~ moved to follow up

### CODE REVIEW

File: provider/core/src/lib.rs [hunk 1]
```diff
@@ -199,13 +199,13 @@ pub mod prelude {
 pub use prelude::*;
 
 // Less important non-prelude items
+pub use crate::any::MaybeSendSync;
```
Comment (sffc) [inline provider/core/src/lib.rs:202]: Issue: If you want to rename the trait, please do this in its own standalone PR so we can merge it before 1.0. Otherwise we should keep the current trait name.

Comment (robertbastian) [reply provider/core/src/lib.rs:202]: I want to do this whole PR before 1.0

File: provider/core/src/response.rs [hunk 2]
```diff
@@ -72,9 +67,16 @@ pub struct DataPayload<M>
 where
     M: DataMarker,
 {
-    pub(crate) yoke: Yoke<M::Yokeable, Option<RcWrap<[u8]>>>,
+    pub(crate) yoke: Yoke<M::Yokeable, Cart>,
 }
 
+#[cfg(feature = "sync")]
+/// The type of the "cart" that is used by `DataPayload`. This changes with the `sync` feature.
+pub type Cart = Option<alloc::sync::Arc<Box<[u8]>>>;
```
Comment (robertbastian) [inline provider/core/src/response.rs:75]: Should we make this `doc(hidden)`?

Comment (Manishearth) [reply provider/core/src/response.rs:75]: I think it's worth publicly mentioning

**Review by Manishearth — APPROVED** [review]
i like this, but i'll wait for shane's review too

**Review by Manishearth — DISMISSED** [review]

File: provider/core/src/response.rs [hunk 3]
```diff
@@ -72,7 +73,45 @@ pub struct DataPayload<M>
...
+impl Deref for Cart {
+    type Target = dyn yoke::erased::ErasedDestructor;
+    fn deref(&self) -> &Self::Target {
+        &*self.0
+    }
+}
+// Safe because both Rc and Arc are CloneableCart
+unsafe impl stable_deref_trait::StableDeref for Cart {}
+// Safe because both Rc and Arc are StableDeref
+unsafe impl yoke::CloneableCart for Cart {}
```
Comment (sffc) [inline provider/core/src/response.rs:95]:
```suggestion
// Safe because both Rc and Arc are StableDeref, and our Deref impl delegates to that Deref impl.
unsafe impl stable_deref_trait::StableDeref for Cart {}
// Safe because both Rc and Arc are CloneableCart, and our Deref impl delegates to that Deref impl.
unsafe impl yoke::CloneableCart for Cart {}
```

File: provider/core/src/response.rs [hunk 4]
```diff
@@ -72,7 +73,45 @@ pub struct DataPayload<M>
...
+        D: 'static + ?Sized,
+    {
+        // Safe because the cart is only wrapped
+        unsafe {
+            Ok(
+                Yoke::try_attach_to_cart(SelectedRc::new(cart), |b: &C| f(&*b))?
+                    .replace_cart(|rc| Cart(rc))
+                    .wrap_cart_in_option(),
+            )
+        }
```
Comment (sffc) [inline provider/core/src/response.rs:113]: Nit (style): Write it such that only the unsafe function is inside the unsafe

```suggestion
        Yoke::try_attach_to_cart(SelectedRc::new(cart), |b: &C| f(&*b))
            // Safe because the cart is only wrapped
            .map(|yoke| unsafe { yoke.replace_cart(|rc| Cart(rc)) })
            .wrap_cart_in_option()
```

File: provider/core/src/response.rs [hunk 5]
```diff
@@ -672,18 +502,28 @@ where
...
-    pub fn from_rc_buffer(buffer: RcWrap<[u8]>) -> Self {
+    /// Converts an owned byte buffer into a `DataPayload<BufferMarker>`.
+    pub fn from_owned_buffer(buffer: Box<[u8]>) -> Self {
         Self {
-            yoke: Yoke::attach_to_cart(buffer, |b| b).wrap_cart_in_option(),
+            // Safe because cart is wrapped
+            yoke: unsafe {
+                Yoke::attach_to_cart(SelectedRc::new(buffer), |b| &**b)
+                    .replace_cart(|b| Some(Cart(b)))
+            },
```
Comment (sffc) [inline provider/core/src/response.rs:512]: Nit: Apply the unsafe only to the unsafe function

File: utils/yoke/src/erased.rs [hunk 6]
```diff
@@ -20,6 +21,12 @@ use alloc::rc::Rc;
 pub trait ErasedDestructor: 'static {}
 impl<T: 'static> ErasedDestructor for T {}
 
+/// A type-erased Cart that has `Arc` semantics
+///
+/// See the docs of [`Yoke::erase_arc_cart()`](crate::Yoke::erase_rc_cart) for more info.
+///
+/// Available with the `"alloc"` feature enabled.
+pub type ErasedArcCart = Arc<dyn ErasedDestructor + Send + Sync>;
```
Comment (sffc) [inline utils/yoke/src/erased.rs:29]: Note: We'll need to release `yoke` if we merge this PR before 1.0 (CC @Manishearth)

File: provider/core/src/response.rs [hunk 7]
```diff
@@ -72,7 +73,45 @@ pub struct DataPayload<M>
...
+    }
+}
+// Safe because both Rc and Arc are CloneableCart
+unsafe impl stable_deref_trait::StableDeref for Cart {}
+// Safe because both Rc and Arc are StableDeref
+unsafe impl yoke::CloneableCart for Cart {}
+
+impl Cart {
+    /// Creates a Yoke<Y, Option<Cart>> from some owned C by applying f.
+    pub fn make_yoke<Y, F, E, C, D>(cart: C, f: F) -> Result<Yoke<Y, Option<Self>>, E>
```
Comment (sffc) [inline provider/core/src/response.rs:99]: Suggestion: `try_make_yoke`

File: provider/core/src/response.rs [hunk 8]
```diff
@@ -52,8 +53,8 @@ pub struct DataResponseMetadata {
 ///
 /// # `sync` feature
 ///
-/// By default, the payload uses an [`Rc<[u8]>`] internally and hence is neither [`Sync`] nor [`Send`].
-/// If these traits are required, the `sync` feature can be enabled to use an [`Arc<[u8]>`] instead.
+/// By default, the payload uses non-concurrent reference counting internally, and hence is neither
+/// [`Sync`] nor [`Send`], if these traits are required, the `sync` feature can be enabled.
```
Comment (sffc) [inline provider/core/src/response.rs:57]: Semicolon Nit

```suggestion
/// By default, the payload uses non-concurrent reference counting internally, and hence is neither
/// [`Sync`] nor [`Send`]; if these traits are required, the `sync` feature can be enabled.
```

**Review by sffc — APPROVED** [review]
So, if I understand this correctly, we could have done this just fine with `RcWrap` and making it wrap a `Box<[u8]>` instead of a `[u8]`. The `Cart` changes are additional cleanups whose advantages include the ability to use `ErasedArcCart` as well as use it internally in BlobDataProvider.

**Comment by sffc** [comment]
Code size: on main:

```
-rwxr-xr-x 1 runner docker    44176 Sep 27 01:49 optim4.elf
-rwxr-xr-x 1 runner docker    27000 Sep 27 01:49 optim5.elf
```

On this branch:

```
-rwxr-xr-x 1 runner docker    43992 Sep 26 20:24 optim4.elf
-rwxr-xr-x 1 runner docker    27144 Sep 26 20:24 optim5.elf
```

Where are the 144 bytes from?

**Comment by robertbastian** [comment]
Code size went up in the last commit only, using the erased yoke. I can undo that, it's an internal change now that we can make later if necessary

**Comment by robertbastian** [comment]
Code size back down to `27000`.

**Comment by robertbastian** [comment]
I've left the Arc changes in yoke for completeness, however we don't need them now and they can just go in the next yoke release.
